### PR TITLE
add yum-plugin-priorites to allow repo override in ci jobs

### PR DESCRIPTION
**Pull request by gabemontero**
See https://github.com/openshift/origin/issues/8571 (though additional work is needed after this merges to resolve that issue)

@smarterclayton @stevekuznetsov @sdodson ptal

### CODE REVIEW

**Comment by smarterclayton** [comment]
Question - should we be installing epel-release if we're going to add our own mirrors?

**Comment by sdodson** [comment]
You're mounting a repo in via secret, but would everyone else?

**Comment by gabemontero** [comment]
@smarterclayton - are you thinking that the Dockerfile say calls a new script we add that varies what `INSTALL_PKGS` is set to based on the presence of our override repo file?

i.e. RUN INSTALL_PKGS=\`newScript.sh\`

And `newScript.sh` echo's a list that either includes epel-release or not based on the presence of the override file.

**Comment by smarterclayton** [comment]
I'm saying whomever builds origin-base decides what mirrors to include, and we only install epel-release if no mirrors are installed (if no mounted mirrors)

**Comment by gabemontero** [comment]
On Tue, Feb 21, 2017 at 5:53 PM Clayton Coleman <notifications@github.com>
wrote:

> I'm saying whomever builds origin-base decides what mirrors to include,
> and we only install epel-release if no mirrors are installed (if no mounted
> mirrors)
>
ok implementation details aside that is what i was inferring.  Thanks for
the confirm - i will rework the pull.

**Comment by gabemontero** [comment]
install of epel-release now conditional ... if local repo file present, epel-release is not included in the list to install

File: images/base/Dockerfile [hunk 1]
```diff
@@ -6,8 +6,16 @@
...
 
-RUN INSTALL_PKGS="which git tar wget hostname sysvinit-tools util-linux bsdtar epel-release \
-      socat ethtool device-mapper iptables tree findutils nmap-ncat e2fsprogs xfsprogs lsof device-mapper-persistent-data ceph-common" && \
+# installation of yum-plugin-priorities must precede installation of epel-release in order for any potential
+# mounted override repo files to take effect
+RUN yum install -y yum-plugin-priorities && \
+    rpm -V yum-plugin-priorities && \
+    yum clean all
+
+RUN if [ -f /etc/yum.repos.d/local_epel.repo ]; then INSTALL_PKGS="which git tar wget hostname sysvinit-tools util-linux bsdtar \
```
Comment (stevekuznetsov) [inline images/base/Dockerfile]: I thought the priority plugin would let us install `epel-release` always? The back we're doing in CI should be transparent to this container manifest.

**Comment by gabemontero** [comment]
On Sat, Mar 4, 2017 at 12:39 PM Steve Kuznetsov <notifications@github.com>
wrote:

> *@stevekuznetsov* commented on this pull request.
> ------------------------------
>
> In images/base/Dockerfile
> <https://github.com/openshift/origin/pull/12979#discussion_r104292957>:
>
> > @@ -6,8 +6,16 @@
>  #
>  FROM centos:centos7
>
> -RUN INSTALL_PKGS="which git tar wget hostname sysvinit-tools util-linux bsdtar epel-release \
> -      socat ethtool device-mapper iptables tree findutils nmap-ncat e2fsprogs xfsprogs lsof device-mapper-persistent-data ceph-common" && \
> +# installation of yum-plugin-priorities must precede installation of epel-release in order for any potential
> +# mounted override repo files to take effect
> +RUN yum install -y yum-plugin-priorities && \
> +    rpm -V yum-plugin-priorities && \
> +    yum clean all
> +
> +RUN if [ -f /etc/yum.repos.d/local_epel.repo ]; then INSTALL_PKGS="which git tar wget hostname sysvinit-tools util-linux bsdtar \
>
> I thought the priority plugin would let us install epel-release always?
> The back we're doing in CI should be transparent to this container manifest.
>
it would but @smarterclayton asked to make it conditional anyway

**Comment by gabemontero** [comment]
[test]

**Comment by stevekuznetsov** [comment]
@smarterclayton what's the rationale for the above? I thought this was supposed to be transparent

**Comment by gabemontero** [comment]
As an aside, our local mirror is no longer working for us (whereas when I prototyped this last month it did work).  Certainly this would need to be resolved before we pull the trigger on this.

Some dependency errors are occurring when running in my test ami (which I provisioned yesterday):

```
--> Finished Dependency Resolution
Error: Package: 1:python-cephfs-0.80.7-0.8.el7.x86_64 (local-epel)
           Requires: python-rados = 1:0.80.7
           Available: 1:python-rados-0.94.5-1.el7.x86_64 (base)
               python-rados = 1:0.94.5-1.el7
Error: Package: 1:ceph-common-0.80.7-0.8.el7.x86_64 (local-epel)
           Requires: python-rbd = 1:0.80.7
           Available: 1:python-rbd-0.94.5-1.el7.x86_64 (base)
               python-rbd = 1:0.94.5-1.el7
Error: Package: 1:ceph-common-0.80.7-0.8.el7.x86_64 (local-epel)
           Requires: python-rados = 1:0.80.7
           Available: 1:python-rados-0.94.5-1.el7.x86_64 (base)
               python-rados = 1:0.94.5-1.el7
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest

```

**Comment by gabemontero** [comment]
I'm still getting the same errors noted in https://github.com/openshift/origin/pull/12979#issuecomment-284843101 

As a reminder, we are simply trying to do a yum install of these packages using our local epel mirror:

`which git tar wget hostname sysvinit-tools util-linux bsdtar socat ethtool device-mapper iptables tree findutils nmap-ncat e2fsprogs xfsprogs lsof device-mapper-persistent-data ceph-common`

@tdawson - @stevekuznetsov believe there is a job that is suppose to periodically update https://mirror.openshift.com/mirror/epel and make sure it is sorted out wrt to error like the one above.

Does that ring any bells, and if so, should we poke that job?  Or is this a known issue? Or if none of the above, any hints on how we should solve / debug?

**Comment by tdawson** [comment]
@gabemontero Is there any way I can get the full error.  This doesn't seem to be an "not all epel is there" issue, but something else.  But I can't tell what from what is shown.

**Comment by gabemontero** [comment]
@tdawson hey - here is a gist with the entire set of output:  https://gist.github.com/gabemontero/cb1cf709ca5164baac4692eacf4923f5  

The vagrant command is executing hack/build-base-images.sh on my ec2 instance.

Also for reference, the update to the Dockerfile that hack/build-base-images.sh uses is captured in this PR.
And the local epel repo file is this:  https://github.com/openshift/vagrant-openshift/blob/master/lib/vagrant-openshift/resources/local_epel.repo

thanks

**Comment by tdawson** [comment]
This isn't a matter of the epel mirror not being in sync.  This is a matter of figuring out why it's trying to pull from local_epel instead of base.  ceph-common, and all it's dependencies, have been merged into the standard RHEL7, and so an install of it should pull everything from base.
To make matters even stranger, you cannot install the ceph-common from epel7.  If you look at it's build https://koji.fedoraproject.org/koji/buildinfo?buildID=738434, for x86_64 it is missing all the packages that we are seeing here.  But only the ppc64le  people complained, so they put it back for them. https://bugzilla.redhat.com/show_bug.cgi?id=1299820
Summary, don't install ceph-common from any epel7 repos.  It just won't work.  Get it from your base repo.

**Comment by gabemontero** [comment]
thanks for the info @tdawson 

The million dollar question for a yum neophyte like me is how to ensure that.

Given the contents of our local_epel.repo:

```
[local-epel]
name=Extra Packages for Enterprise Linux 7 - $basearch
baseurl=https://mirror.openshift.com/mirror/epel/7/x86_64
#mirrorlist=https://mirrors.fedoraproject.org/metalink?repo=epel-7&arch=$basearch
failovermethod=priority
enabled=1
gpgcheck=1
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7
# according to note from sdodson in https://github.com/openshift/origin/issues/8571, default priority is 99,
# but to ensure our local mirror is not overriden, setitng to 1
priority=1

[local-epel-debuginfo]
name=Extra Packages for Enterprise Linux 7 - $basearch - Debug
baseurl=https://mirror.openshift.com/mirror/epel/7/x86_64/debug
#mirrorlist=https://mirrors.fedoraproject.org/metalink?repo=epel-debug-7&arch=$basearch
failovermethod=priority
enabled=0
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7
gpgcheck=1
# according to note from sdodson in https://github.com/openshift/origin/issues/8571, default priority is 99,
# but to ensure our local mirror is not overriden, setitng to 1
priority=1
```

And the contents of the epel.repo that came with the system:

```
[epel]
name=Extra Packages for Enterprise Linux 7 - $basearch
#baseurl=http://download.fedoraproject.org/pub/epel/7/$basearch
mirrorlist=https://mirrors.fedoraproject.org/metalink?repo=epel-7&arch=$basearch
failovermethod=priority
enabled=1
gpgcheck=1
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7

[epel-debuginfo]
name=Extra Packages for Enterprise Linux 7 - $basearch - Debug
#baseurl=http://download.fedoraproject.org/pub/epel/7/$basearch/debug
mirrorlist=https://mirrors.fedoraproject.org/metalink?repo=epel-debug-7&arch=$basearch
failovermethod=priority
enabled=0
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7
gpgcheck=1

[epel-source]
name=Extra Packages for Enterprise Linux 7 - $basearch - Source
#baseurl=http://download.fedoraproject.org/pub/epel/7/SRPMS
mirrorlist=https://mirrors.fedoraproject.org/metalink?repo=epel-source-7&arch=$basearch
failovermethod=priority
enabled=0
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7
gpgcheck=1
```

I have no idea how to enforce that.  So far, I've tried to
1) revert the priority back to 98 in local_epel.repo ... same error
2) switch from baseurl to mirrorlist  ... get an erorr msg basically saying need baseurl ... might be a priority plugin thing

I could remove the priority from there, but then it seems like I need to remove the epel.repo file, which we'd rather not do.

Any suggestions?

**Comment by tdawson** [comment]
Is there any way I/we can look at what the [base] repo config looks like.
I just want to make sure that [base] has priority set correct and it's pointing to a repo that has the packages we want.

**Comment by gabemontero** [comment]
Hey @tdawson - sorry for the delay - was out a bit for surgery, and am now finally about caught up.

Interestingly, when looking at all the files in `/etc/yum.repos.d` on the ec2 instance we are testing on, none of the `*.repo` files have `[base]` in them.

Is there another place in our instance I should look for that?  Or of course perhaps we need to inject an appropriate `[base]` repo file when employing our local_epel.repo and the openshift mirror ?

**Comment by gabemontero** [comment]
Couldn't install the yum plugin priorities with the last test run - none of the mirrors had it.

**Comment by bparees** [comment]
is this PR still being pursued?

**Comment by gabemontero** [comment]
Given https://github.com/openshift/origin/pull/16028 if the removal of epel goes through I think this PR and the corresponding flake issue can get nuked 

@stevekuznetsov - agree ?

**Pull request closed without merging**